### PR TITLE
feat: Out of the box metrics for resource manager

### PR DESCRIPTION
**Pull request by MarcoPolo**
This adds metrics for resource manager that allows an end user to see what's happening with their go-libp2p application.

This hooks into the existing tracing setup to emit metrics and aggregate them when appropriate.

Possible improvements:
1. Record limits

Example graphs:

Note the histogram is not rendered correctly in this snapshot. See the screenshot below for how it should look.
https://snapshots-origin.grafana.net/dashboard/snapshot/nDCjmTCnHpmEpXgtcWO7ThEDS7B5DLLn?orgId=2

How the histogram should look like:
Screen Shot 2022-06-20 at 5 46 49 PM

I'll see if I can spin up a public ipfs dashboard to demo.

### CODE REVIEW

File: obs/stats.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+var (
+	SystemOutboundConnsView = &view.View{Measure: systemOutboundConns, Aggregation: view.Sum()}
+	SystemInboundConnsView  = &view.View{Measure: systemInboundConns, Aggregation: view.Sum()}
+
+	ConnView = &view.View{Measure: conns, Aggregation: view.Sum(), TagKeys: []tag.Key{Direction, Scope}}
+
+	fibLikeDistribution = []float64{
+		1.1, 2.1, 3.1, 5.1, 8.1, 13.1, 21.1, 34.1, 55.1, 100.1, 200.1,
+	}
```
Comment (mxinden) [inline obs/stats.go:59]: Out of curiosity, why this distribution?

**Comment by marten-seemann** [comment]
Looking at the dashboard, the p90 values are below the p50 values. That doesn't seem correct, or am I missing something?

File: obs/stats.go (hunk 1 above)
Comment (MarcoPolo) [reply obs/stats.go:59]: fibonaccci with +0.1 because of a quirk in opencensus that does bound non-inclusive bound checking: https://github.com/census-instrumentation/opencensus-go/blob/v0.23.0/stats/view/aggregation_data.go#L195.

But the fib sequence itself is totally arbitrary. Happy to change this to something else if you have a better idea.

**Comment by MarcoPolo** [comment]
Public dashboard: https://ipfs.marcopolo.io/d/MgmGIjjnk/resource-manager?orgId=1

**Comment by MarcoPolo** [comment]
> Looking at the dashboard, the p90 values are below the p50 values. That doesn't seem correct, or am I missing something?

Resolved out of band. The aggregated number of peers was misinterpreted. I changed the color to make it more obvious this isn't p50

File: obs/stats.go (hunk 1 above)
Comment (mxinden) [reply obs/stats.go:59]: No better ideas. The go-to distributions are _linear_ and _exponential_ in the Prometheus ecosystem:

https://github.com/prometheus/client_golang/blob/4ad265f1b4ee8d5c239b3f36081529bbe3a838db/prometheus/histogram.go#L75-L145

**Review by vyzo — COMMENTED** [review]
Can we make sure this is optional, enabled with a constructor option?
We shouldn't be forcing down the throat of developers any metrics solution.

Note that with regards to UX, this can be easily aggregated in a go-libp2p constructor metrics option, which initializes various components (including the rcmgr) with metrics.

**Comment by MarcoPolo** [comment]
> Can we make sure this is optional, enabled with a constructor option?
> We shouldn't be forcing down the throat of developers any metrics solution.

Yep. There's little value to these if they aren't hooked up to an exporter like prometheus.

> Note that with regards to UX, this can be easily aggregated in a go-libp2p constructor metrics option, which initializes various components (including the rcmgr) with metrics.

Good idea!

**Comment by MarcoPolo** [comment]
I tried getting exemplars through, but while you can attach an exemplar via opencensus the opencensus prometheus exporter doesn't support emitting it.

File: rcmgr.go [hunk 2]
```diff
@@ -423,13 +424,93 @@ func newStreamScope(dir network.Direction, limit Limit, peer *peerScope, rcmgr *
...
-			fmt.Sprintf("stream-%d", rcmgr.nextStreamId()), rcmgr.trace, rcmgr.metrics),
+			streamScopeName(rcmgr.nextStreamId()), rcmgr.trace, rcmgr.metrics),
 		dir:   dir,
 		rcmgr: peer.rcmgr,
 		peer:  peer,
 	}
 }
 
+func IsSystemScope(name string) bool {
+	return strings.HasPrefix(name, "system")
```
Comment (marten-seemann) [inline rcmgr.go:435]: This would return `true` for a span in the system scope. Is that intended?

File: obs/stats.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+	PeerConnsNegativeView = &view.View{
+		Measure:     peerConnsNegative,
+		Aggregation: view.Distribution(oneTenThenExpDistribution...),
+		TagKeys:     []tag.Key{directionTag},
+	}
+
+	StreamView             = &view.View{Measure: streams, Aggregation: view.Sum(), TagKeys: []tag.Key{directionTag, scopeTag, serviceTag, protocolTag}}
+	PeerStreamsView        = &view.View{Measure: peerStreams, Aggregation: view.Distribution(oneTenThenExpDistribution...), TagKeys: []tag.Key{directionTag}}
+	PeerStreamNegativeView = &view.View{Measure: peerStreamsNegative, Aggregation: view.Distribution(oneTenThenExpDistribution...), TagKeys: []tag.Key{directionTag}}
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply obs/stats.go:67]: I think so. From https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations

> In principle, however, you can use summaries and histograms to observe negative values (e.g. temperatures in centigrade). In that case, the sum of observations can go down, so you cannot apply rate() to it anymore. In those rare cases where you need to apply rate() and cannot avoid negative observations, you can use two separate summaries, one for positive and one for negative observations (the latter with inverted sign), and combine the results later with suitable PromQL expressions.

Our metrics don't have a notion of removing a previous observation. The best we can do is add a new negative observation that subtracts the previous observation, then add a new observation.

So we need to make two observations:
1. We want to record the old number of streams this peer _had_ (this is negative).
2. We want to record the current number of streams this peer has.

In order to make a negative observation with histograms we need to record it in a separate measurement, then combine it later.

That's my understanding, but it's possible I'm misunderstanding something. @mxinden does the above seem correct to you?

File: rcmgr.go (hunk 2 above)
Comment (MarcoPolo) [reply rcmgr.go:435]: Not indended, will fix

File: obs/stats.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mxinden) [reply obs/stats.go:67]: > Our metrics don't have a notion of removing a previous observation.

I am not familiar enough with OpenCensus, thus I am not sure what their best practice is.

> From https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations
> 
> > In principle, however, you can use summaries and histograms to observe negative values (e.g. temperatures in centigrade). In that case, the sum of observations can go down, so you cannot apply rate() to it anymore. In those rare cases where you need to apply rate() and cannot avoid negative observations, you can use two separate summaries, one for positive and one for negative observations (the latter with inverted sign), and combine the results later with suitable PromQL expressions.

I don't think this is referring to _undoing_ a previous observation, but only to record both negative and positive observations. That is not to say that your trick here won't work. Mind sharing the queries you are using here? Are you actually using the `histogram_quantile` function? Note that `histogram_quantile` calculates the quantiles over time, thus I am not sure the above _undo_ logic is compatible with it.

To come back to the original goal, do I understand correctly that you would like to answer the following question:

_How many currently connected peers have 1, 10, 100, 1000, 10000 (arbitrary distribution) streams?_

In case you only want to visualize the above with a bar chart, but not calculate quantiles, how about exposing a Prometheus Gauge for each of these buckets. You can `inc` and `dec` a Prometheus Gauge.

[the remark this answers is not in the document]
Comment (MarcoPolo) [reply obs/stats.go:67]: > Note that histogram_quantile calculates the quantiles over time

To be clear `https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile` uses an instant-vector so it calculates the quantile at each point in time. As opposed to something like `rate` which uses a range vector that can look at multiple timestamps for each point. (I'm sure you know this, just adding it here for completeness).

> Mind sharing the queries you are using here? Are you actually using the histogram_quantile function?

```
histogram_quantile(0.90, (rcmgr_trace_metrics_peer_streams_bucket - rcmgr_trace_metrics_peer_streams_negative_bucket))
```

This is similar to the commonly used [example](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile): `histogram_quantile(0.9, rate(http_request_duration_seconds_bucket[10m]))`. Except instead of using rate to get the rate of increase (by taking the current value and comparing it with a previous value) and calculating quantiles from that I'm using these two histograms to create a histogram of the current state and calculating quantiles with that.

Let me rephrase:
1. We agree this example is correct: `histogram_quantile(0.9, rate(http_request_duration_seconds_bucket[10m]))`. Since it's from the docs.
2. This example calculates quantiles on a new histogram. This new histogram is generated by doing `http_request_duration_seconds_bucket - http_request_duration_seconds_bucket[previous_time]` with the `rate` function (https://github.com/prometheus/prometheus/blob/main/promql/functions.go#L74)
3. What we're doing is keeping track of our last observation (the negative bucket), and generating a new histogram with our current state of the world with `rcmgr_trace_metrics_peer_streams_bucket - rcmgr_trace_metrics_peer_streams_bucket@previous_observation (aka negative bucket)`. Which I'm arguing is the same as 2, except we are explicit in what our previous observation was.
  a. Aside, we could use rate here as well and only use the one bucket, but then we'd be only seeing the differences of when a peer moves to a different bucket which is not as useful as seeing the current distribution of streams per peer. There might be a way with promql to do what I want, but I haven't figured it out (essentially what's the difference between the current value and the last different value).
  

Or another way to think about this is:
1. Assume `rcmgr_trace_metrics_peer_streams_bucket - rcmgr_trace_metrics_peer_streams_negative_bucket` generates a valid histogram, let's call it `H`.
2. We should be able to caluclate the quantiles from this histogram `H` with the `histogram_quantile` function. (Which is actually quite simple: https://github.com/prometheus/prometheus/blob/main/promql/quantile.go#L101.)

So the question becomes, is our assumption of 1 here correct? I'm arguing it is because it is the histogram of our current state of streams per peer. It also looks exactly like a histogram metric promql would expect (cumulative le distributions).

I know that's quite a lot, and I appreciate your input here! Thank you for thinking about this and checking my work :)

[the remark this answers is not in the document]
Comment (mxinden) [reply obs/stats.go:67]: > > Note that histogram_quantile calculates the quantiles over time
> 
> To be clear `https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile` uses an instant-vector so it calculates the quantile at each point in time. As opposed to something like `rate` which uses a range vector that can look at multiple timestamps for each point. (I'm sure you know this, just adding it here for completeness).
> 
> > Mind sharing the queries you are using here? Are you actually using the histogram_quantile function?
> 
> ```
> histogram_quantile(0.90, (rcmgr_trace_metrics_peer_streams_bucket - rcmgr_trace_metrics_peer_streams_negative_bucket))
> ```
> 
> This is similar to the commonly used [example](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile): `histogram_quantile(0.9, rate(http_request_duration_seconds_bucket[10m]))`. Except instead of using rate to get the rate of increase (by taking the current value and comparing it with a previous value) and calculating quantiles from that I'm using these two histograms to create a histogram of the current state and calculating quantiles with that.

This is really clever! To be honest, I didn't think about not using `rate` with a `histogram_quantile`. Thanks for the explainer! :bow:

> 3\. Aside, we could use rate here as well and only use the one bucket, but then we'd be only seeing the differences of when a peer moves to a different bucket which is not as useful as seeing the current distribution of streams per peer.

:+1: 

> 3\. There might be a way with promql to do what I want, but I haven't figured it out (essentially what's the difference between the current value and the last different value).

Isn't this impossible at the data level as one would need to know which decrease in one bucket matches a corresponding increase in another bucket and which increase is a new peer, thus does not have a decrease in another bucket?

> @mxinden does the above seem correct to you?

The above seems correct to me!

In case you can influence the Prometheus metric `# HELP` text, I think it is worth a mention of the concept you are using here.

**Review by marten-seemann — APPROVED** [review]
CI is still unhappy about this PR though.

**Comment by MarcoPolo** [comment]
(rebased off master to fix conflict)
